### ISSUE TITLE
Images are not rendered after application resumes from background

### ISSUE THREAD

**Issue by ezamagni** [flutter/flutter#64558]
## Steps to Reproduce

This issue only affects our iOS builds.

1. Create and run an application that displays some images
2. Suspend the app and leave it in background for a while (usually a matter of an hour or more)
3. (Optional: not sure it is related to the issue) Have the application perform some background task
4. Relaunch the app

**Expected results:** 
The Image widgets displayed in step 1 should still be visible and fully functioning.

**Actual results:** 
Most of all of the Image widgets that were on-screen when the app went in background are no longer displaying their image content after app resumes.

## Details

- Both images downloaded via network (cached by [cached_network_image](https://pub.dev/packages/cached_network_image)) and _some_ assets embedded in application are missing after app resumes.
- PNGs used for small icons seems to be unaffected: in the attached screenshot the dark-grey (hard to see) picture at the center of screen is the only missing non-network asset and it happens to be a JPEG (not sure if related).
- all subsequent image renderings _after_ the app resumes are fully functional
- this application was also suffering from [#55969](https://github.com/flutter/flutter/issues/55969), we suspect this might be a related regression or some remaining metal rendering issue.

![IMG_3064](https://user-images.githubusercontent.com/6697707/91178777-ee695080-e6e5-11ea-8dbf-f1b5c874ead7.jpg)

![IMG_3069](https://user-images.githubusercontent.com/6697707/91178981-36887300-e6e6-11ea-9aa6-e699fdab0024.jpg)

<details>
  <summary>Logs</summary>

Our .ipa is built in CI using Bitrise. The following log comes from the virtual machine that executes our build workflow.

```
+ flutter doctor -v
Downloading Material fonts...                                       0.5s
Downloading Gradle Wrapper...                                       0.1s
Downloading package sky_engine...                                   0.2s
Downloading flutter_patched_sdk tools...                            0.9s
Downloading flutter_patched_sdk_product tools...                    0.8s
Downloading darwin-x64 tools...                                     1.3s
Downloading libimobiledevice...                                     0.1s
Downloading usbmuxd...                                              0.1s
Downloading libplist...                                             0.1s
Downloading openssl...                                              0.2s
Downloading ios-deploy...                                           0.1s
Downloading darwin-x64/font-subset tools...                         0.2s
Downloading android-arm-profile/darwin-x64 tools...                 0.4s
Downloading android-arm-release/darwin-x64 tools...                 0.2s
Downloading android-arm64-profile/darwin-x64 tools...               0.4s
Downloading android-arm64-release/darwin-x64 tools...               0.4s
Downloading android-x64-profile/darwin-x64 tools...                 0.2s
Downloading android-x64-release/darwin-x64 tools...                 0.2s
[✓] Flutter (Channel stable, 1.20.2, on Mac OS X 10.15.6 19G73, locale en-US)
    • Flutter version 1.20.2 at /Users/vagrant/flutter-sdk/flutter
    • Framework revision bbfbf1770c (12 days ago), 2020-08-13 08:33:09 -0700
    • Engine revision 9d5b21729f
    • Dart version 2.9.1
[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.0)
    • Android SDK at /usr/local/share/android-sdk
    • Platform android-30, build-tools 30.0.0
    • ANDROID_HOME = /usr/local/share/android-sdk
    • ANDROID_SDK_ROOT = /usr/local/share/android-sdk
    • Java binary at: /Library/Java/JavaVirtualMachines/adoptopenjdk-8.jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (AdoptOpenJDK)(build 1.8.0_252-b09)
    • All Android licenses accepted.
[✓] Xcode - develop for iOS and macOS (Xcode 11.6)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 11.6, Build version 11E708
    • CocoaPods version 1.9.1
[!] Android Studio (not installed)
    • Android Studio not found; download from https://developer.android.com/studio/index.html
      (or visit https://flutter.dev/docs/get-started/install/macos#android-setup for detailed instructions).
[!] Connected device
    ! No devices available
! Doctor found issues in 2 categories.
```

</details>

**Comment by TahaTesser** [comment]
Hi @ezamagni 
Does this happen if the images are in listview or rows? The layout doesn' matter? 
Can you please provide [minimal complete reproducible code sample](https://stackoverflow.com/help/minimal-reproducible-example) with sample images from Unsplash to better address the issue 
Thank you

**Comment by aaronscherbing** [comment]
We're seeing this as well.  It could potentially be related to the following issue: #57676 

All of the missing imagery in the attached image is an `Image` widget using a custom `ImageProvider` that makes use of a library called `flutter_cache_manager` to initiate the image fetching. (ie. the image provider reads from the cache first, and if there's a miss, it makes an http request for the image.)

As for your layout questions @TahaTesser, most of the missing images in the attached screenshot are inside CustomScrollViews (horizontal and vertical).  The image in the bottom right corner is part of a Row, which is a Scaffold's bottomNavigationBar.

![91345939-c5a89000-e7a5-11ea-82f0-f5f8146c2531](https://user-images.githubusercontent.com/11944082/91440126-bcfe9b00-e833-11ea-8351-3c8cc377a32a.png)

**Comment by aaronscherbing** [comment]
A few more things that might help with troubleshooting:

We don't see this issue in Flutter version 1.15.4.  That's the version we're releasing with currently.  All the builds using Flutter versions after the transition to Metal on iOS have had rendering issues related to the app being in the background for a period of time.

We don't see the issue on any of our Android builds.

**Comment by ezamagni** [comment]
Unfortunately, i'm having quite a hard time coming up with a minimal reproducible sample. I will continue experimenting and will post something as soon as possible.

Regarding the layout questions: we're seeing this issue happening both in widgets nested inside complex layouts such as `NestedScrollView` -> `CustomScrollView` trees, as well as in simple cases such as `BottomNavigationBar` -> `BottomNavigationBarItem` (see the user thumbnail in the bottom bar, in my previous screenshots).

**Comment by TahaTesser** [comment]
@ezamagni 
Are you using the same`flutter_cache_manager` and `ImageProvider`? 

@aaronscherbing  @ezamagni 
To better address the issue, can you please provide your `flutter doctor -v` and a complete reproducible minimal code sample

I have attempted to reproduce this issue with a simple app with jpg images. After 2 hours of keeping the app in the background. Images are rendered properly on the latest `stable` 

<details> 
<summary>code sample</summary> 

```dart
import 'package:flutter/material.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Material App',
      theme: ThemeData.dark(),
      home: Home(),
    );
  }
}

class Home extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    final List<String> _tabs = ['Tab 1', 'Tab 2'];

    return Scaffold(
      body: DefaultTabController(
          length: _tabs.length,
          child: NestedScrollView(
            headerSliverBuilder:
                (BuildContext context, bool innerBoxIsScrolled) {
              return <Widget>[
                SliverOverlapAbsorber(
                  handle:
                      NestedScrollView.sliverOverlapAbsorberHandleFor(context),
                  sliver: SliverAppBar(
                    title: const Text('Images are not rendered'),
                    pinned: true,
                    expandedHeight: 150.0,
                    forceElevated: innerBoxIsScrolled,
                    bottom: TabBar(
                      tabs:
                          _tabs.map((String name) => Tab(text: name)).toList(),
                    ),
                  ),
                ),
              ];
            },
            body: TabBarView(
              children: _tabs.map((String name) {
                return SafeArea(
                  top: false,
                  bottom: false,
                  child: Builder(
                    // This Builder is needed to provide a BuildContext that is
                    // "inside" the NestedScrollView, so that
                    // sliverOverlapAbsorberHandleFor() can find the
                    // NestedScrollView.
                    builder: (BuildContext context) {
                      return CustomScrollView(
                        // The "controller" and "primary" members should be left
                        // unset, so that the NestedScrollView can control this
                        // inner scroll view.
                        // If the "controller" property is set, then this scroll
                        // view will not be associated with the NestedScrollView.
                        // The PageStorageKey should be unique to this ScrollView;
                        // it allows the list to remember its scroll position when
                        // the tab view is not on the screen.
                        key: PageStorageKey<String>(name),
                        slivers: <Widget>[
                          SliverOverlapInjector(
                            handle:
                                NestedScrollView.sliverOverlapAbsorberHandleFor(
                                    context),
                          ),
                          SliverPadding(
                            padding: const EdgeInsets.all(8.0),
                            sliver: SliverList(
                              delegate: SliverChildBuilderDelegate(
                                (BuildContext context, int index) {
                                  print('assets/image0\$i.png');
                                  return Container(
                                    margin: const EdgeInsets.all(10),
                                    width: double.infinity,
                                    height: 500,
                                    child: Image.asset(
                                      'assets/image0${index + 1}.jpg',
                                      fit: BoxFit.cover,
                                    ),
                                  );
                                },
                                childCount: 6,
                              ),
                            ),
                          ),
                        ],
                      );
                    },
                  ),
                );
              }).toList(),
            ),
          )),
      floatingActionButton: FloatingActionButton(
        child: Icon(Icons.add),
        onPressed: () {},
      ),
    );
  }
}

```

</details> 

<details> 
<summary>flutter doctor -v</summary> 

```
[✓] Flutter (Channel stable, 1.20.2, on Mac OS X 10.15.6 19G2021, locale en-GB)
    • Flutter version 1.20.2 at /Users/tahatesser/Code/flutter_stable
    • Framework revision bbfbf1770c (2 weeks ago), 2020-08-13 08:33:09 -0700
    • Engine revision 9d5b21729f
    • Dart version 2.9.1

 
[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.1)
    • Android SDK at /Users/tahatesser/Code/sdk
    • Platform android-30, build-tools 30.0.1
    • ANDROID_HOME = /Users/tahatesser/Code/sdk
    • Java binary at: /Applications/Android
      Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6222593)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 11.6)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 11.6, Build version 11E708
    • CocoaPods version 1.9.3

[✓] Android Studio (version 4.0)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 48.1.2
    • Dart plugin version 193.7547
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6222593)

[✓] VS Code (version 1.48.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.13.2

[!] Connected device
    ! No devices available

! Doctor found issues in 1 category.

```

</details>

**Comment by brianb-sf** [comment]
To add some more context, this only seems to happen when the application is woken up while in the background. In our case, when the user allows background location permissions and enters a predefined geofence then brings the application into the foreground this issue reproduces. It also seems to only happen in release mode. 

Not sure how to provide a code sample with geofences set up that could reproduce this for others. I literally have to drive 30 mins from one geofence to another to get it to happen. The important bit is that the application has to we woken up to perform some work in the background.

**Comment by mahdi-ninja** [comment]
We have the same issue as well. We are also using cached_network_image to display images and we also use flutter_background_geolocation to perform background tasks.

**Comment by AhmedKhattab95** [comment]
@mahdi-ninja is this also happen in flutter 1.20.2 with you?

**Comment by mahdi-ninja** [comment]
> @mahdi-ninja is this also happen in flutter 1.20.2 with you?

Yes. It happens on 1.20.2.

**Comment by TahaTesser** [comment]
I would love to reproduce this even if takes few hours  @mahdi-ninja  
Can you please provide a complete reproducible minimal code sample
Thank you

**Comment by ezamagni** [comment]
I managed to reproduce this issue 2 times in the timespan of ~24 hr with a very simple app:

Here is the dart code: a list with 3 images, one is loaded from the network, the second one is cached through cached_network_image, the third is an asset (please provide a jpg in the path `img/beach.jpg`, and  register it in the .yaml).
<details>
  <summary>main.dart</summary>

```dart
import 'package:cached_network_image/cached_network_image.dart';
import 'package:flutter/material.dart';


void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        brightness: Brightness.dark,
        primarySwatch: Colors.blue,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: MyHomePage(),
    );
  }
}

class MyHomePage extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return SafeArea(
      top: false,
      child: CustomScrollView(
        slivers: [
          SliverToBoxAdapter(child: Container(height: 24)),
          SliverToBoxAdapter(
            child: SimpleCard(
              child: Image.network(
                'https://images.unsplash.com/photo-1458966480358-a0ac42de0a7a?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=1650&q=80',
                fit: BoxFit.fitWidth,
              ),
            ),
          ),
          SliverToBoxAdapter(child: Container(height: 12)),
          SliverToBoxAdapter(
            child: SimpleCard(
              child: CachedNetworkImage(
                imageUrl:
                    "https://images.unsplash.com/photo-1502082553048-f009c37129b9?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=2250&q=80",
                placeholder: (context, url) => Center(
                  child: CircularProgressIndicator(),
                ),
                fit: BoxFit.fitWidth,
                errorWidget: (context, url, error) => Icon(Icons.error),
              ),
            ),
          ),
          SliverToBoxAdapter(child: Container(height: 12)),
          SliverToBoxAdapter(
            child: SimpleCard(
              child: Image.asset(
                "img/beach.jpg",
                fit: BoxFit.fitWidth,
              ),
            ),
          ),
          SliverToBoxAdapter(child: Container(height: 24)),
        ],
      ),
    );
  }
}

class SimpleCard extends StatelessWidget {
  final Widget child;

  SimpleCard({
    this.child,
    Key key,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return AspectRatio(
      aspectRatio: 16 / 10,
      child: ClipRRect(
        borderRadius: BorderRadius.circular(12),
        child: Container(
          color: Colors.blueGrey,
          child: child,
        ),
      ),
    );
  }
}
```
</details>

Here, i'm implementing a callback in the AppDelegate to have iOS (hopefully) assign some background execution time to perform fetch operations (it doesn't matter if the call completes successfully).

<details>
<summary>AppDelegate.swift</summary>

```swift
@objc class AppDelegate: FlutterAppDelegate {
    
    override func application(_ application: UIApplication, didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?) -> Bool {
        GeneratedPluginRegistrant.register(with: self)
        UIApplication.shared.setMinimumBackgroundFetchInterval(30 * 60)
        
        return super.application(application, didFinishLaunchingWithOptions: launchOptions)
    }
    
    override func application(
        _ application: UIApplication,
        performFetchWithCompletionHandler completionHandler: @escaping (UIBackgroundFetchResult) -> Void
    ) {
        let logger = OSLog(subsystem: Bundle.main.infoDictionary!["CFBundleIdentifier"] as! String, category: "AppDelegate")
        os_log("performing background fetch", log: logger, type: .info)
        let url = URL(string: "https://ene76rx5st1xk.x.pipedream.net")!
        let urlSession = URLSession(configuration: URLSessionConfiguration.default)
        let task = urlSession.dataTask(with: URLRequest(url: url)) { (data, _, error) in
            if let error = error {
                os_log("background fetch failed: %{public}@", log: logger, type: .error, String(describing: error))
                completionHandler(.failed)
            } else {
                os_log("background fetch completed", log: logger, type: .info)
                completionHandler(.newData)
            }
        }
        task.resume()
    }
}
```
Remember to opt-in to the background fetch capability inside the target settings.
<img width="387" alt="Schermata 2020-09-02 alle 14 25 41" src="https://user-images.githubusercontent.com/6697707/91980914-504e3980-ed28-11ea-9cd3-bc4dea5349bb.png">

</details>

Once the app is ready, build it in release mode and keep it installed on your device, without any debugger attached to it.
Now the difficult part is actually triggering the issue: usually i open the app from time to time, while casually using other app, or have the device rest for a few hours.

This is a video demonstrating what happens once the issue is triggered: one of the image is gone.
Note what happens when the device switches in landscape and then back in portrait: the image gets rendered correctly.
[RPReplay_Final1598976989.mov.zip](https://github.com/flutter/flutter/files/5162321/RPReplay_Final1598976989.mov.zip)

**Comment by csells** [comment]
@goderbauer is this related to the state restorage work you've been doing?

**Comment by TahaTesser** [comment]
I can  finally reproduce the issue after several hours of installing the app based on https://github.com/flutter/flutter/issues/64558#issuecomment-685707049 steps and code sample 

Here what I did 
Using the code sample and assets

<details> 
<summary>main.dart</summary> 

```dart
import 'package:cached_network_image/cached_network_image.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        brightness: Brightness.dark,
        primarySwatch: Colors.blue,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: MyHomePage(),
    );
  }
}

class MyHomePage extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return SafeArea(
      top: false,
      child: CustomScrollView(
        slivers: [
          SliverToBoxAdapter(child: Container(height: 24)),
          SliverToBoxAdapter(
            child: SimpleCard(
              child: Image.network(
                'https://images.unsplash.com/photo-1458966480358-a0ac42de0a7a?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=1650&q=80',
                fit: BoxFit.fitWidth,
              ),
            ),
          ),
          SliverToBoxAdapter(child: Container(height: 12)),
          SliverToBoxAdapter(
            child: SimpleCard(
              child: CachedNetworkImage(
                imageUrl:
                    "https://images.unsplash.com/photo-1502082553048-f009c37129b9?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=2250&q=80",
                placeholder: (context, url) => Center(
                  child: CircularProgressIndicator(),
                ),
                fit: BoxFit.fitWidth,
                errorWidget: (context, url, error) => Icon(Icons.error),
              ),
            ),
          ),
          SliverToBoxAdapter(child: Container(height: 12)),
          SliverToBoxAdapter(
            child: SimpleCard(
              child: Image.asset(
                "img/beach.jpg",
                fit: BoxFit.fitWidth,
              ),
            ),
          ),
          SliverToBoxAdapter(child: Container(height: 24)),
        ],
      ),
    );
  }
}

class SimpleCard extends StatelessWidget {
  final Widget child;

  SimpleCard({
    this.child,
    Key key,
  }) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return AspectRatio(
      aspectRatio: 16 / 10,
      child: ClipRRect(
        borderRadius: BorderRadius.circular(12),
        child: Container(
          color: Colors.blueGrey,
          child: child,
        ),
      ),
    );
  }
}

```

</details> 

<details> 
<summary>AppDelegate.swift</summary> 

```swift
import UIKit
import Flutter
import os

@UIApplicationMain
@objc class AppDelegate: FlutterAppDelegate {
  override func application(
    _ application: UIApplication,
    didFinishLaunchingWithOptions launchOptions: [UIApplication.LaunchOptionsKey: Any]?
  ) -> Bool {
    GeneratedPluginRegistrant.register(with: self)
    UIApplication.shared.setMinimumBackgroundFetchInterval(30 * 60)
    return super.application(application, didFinishLaunchingWithOptions: launchOptions)
  }
    
    override func application(
        _ application: UIApplication,
        performFetchWithCompletionHandler completionHandler: @escaping (UIBackgroundFetchResult) -> Void
    ) {
        let logger = OSLog(subsystem: Bundle.main.infoDictionary!["CFBundleIdentifier"] as! String, category: "AppDelegate")
        os_log("performing background fetch", log: logger, type: .info)
        let url = URL(string: "https://ene76rx5st1xk.x.pipedream.net")!
        let urlSession = URLSession(configuration: URLSessionConfiguration.default)
        let task = urlSession.dataTask(with: URLRequest(url: url)) { (data, _, error) in
            if let error = error {
                os_log("background fetch failed: %{public}@", log: logger, type: .error, String(describing: error))
                completionHandler(.failed)
            } else {
                os_log("background fetch completed", log: logger, type: .info)
                completionHandler(.newData)
            }
        }
        task.resume()
    }
}
```

</details> 

**image asset**

[img.zip](https://github.com/flutter/flutter/files/5168100/img.zip)



Executed `flutter build ios --release` and then `flutter install` to iPhone XR

Occasionally opened the flutter app, switched other apps, played some Youtube videos  for a couple of hours and when I switched to flutter same results as https://github.com/flutter/flutter/issues/64558#issuecomment-685707049

**First Run**
![IMG_0128](https://user-images.githubusercontent.com/48603081/92107614-b8258280-ee03-11ea-8959-8261af4881a6.PNG)


**After App Resumes from background**
![IMG_0130](https://user-images.githubusercontent.com/48603081/92107616-b9ef4600-ee03-11ea-9f3e-be9a1eba796f.PNG)


 

<details> 
<summary>flutter doctor -v</summary> 

```
[✓] Flutter (Channel stable, 1.20.3, on Mac OS X 10.15.6 19G2021, locale en-GB)
    • Flutter version 1.20.3 at /Users/tahatesser/Code/flutter_stable
    • Framework revision 216dee60c0 (2 days ago), 2020-09-01 12:24:47 -0700
    • Engine revision d1bc06f032
    • Dart version 2.9.2

 
[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
    • Android SDK at /Users/tahatesser/Code/sdk
    • Platform android-29, build-tools 29.0.2
    • ANDROID_HOME = /Users/tahatesser/Code/sdk
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 11.7)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 11.7, Build version 11E801a
    • CocoaPods version 1.9.3

[✓] Android Studio (version 4.0)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 48.1.2
    • Dart plugin version 193.7547
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)

[✓] VS Code (version 1.48.2)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.13.2

[✓] Connected device (1 available)
    • iPhone 11 (mobile) • EDB19945-0B9F-4A6E-BAA1-425E68E3B6D6 • ios •
      com.apple.CoreSimulator.SimRuntime.iOS-13-7 (simulator)

• No issues found!
```

</details>

**Comment by HansMuller** [comment]
CC @dnfield

**Comment by dnfield** [comment]
@chinmaygarde spent some time looking at resumption from background tasks and Metal and might know what's going on here. 

It seems strange that only the network image is failing to load though. I wonder if there's anything in the console when this happens - perhaps iOS doesn't like us trying to open a raw socket in the background, if in fact we're trying to.

**Comment by ezamagni** [comment]
If it's worth something, we are experiencing this issue even on flutter channel beta (1.21.0-9.2.pre).

@dnfield  i agree, it's very strange that only Image.network is affected: in our application, almost every kind of image asset is affected. Maybe the code sample i provided is not well suited to trigger the Metal issue.

Is there any ETA on when this might get addressed, even if on a dev channel?
This issue is becoming _very_ problematic for us.

**Comment by dnfield** [comment]
Are you able to trigger this by doing a launch from background in Xcode?

**Comment by ezamagni** [comment]
> Are you able to trigger this by doing a launch from background in Xcode?

Unfortunately, i've never been able to. As @brianb-sf reported in [this comment](https://github.com/flutter/flutter/issues/64558#issuecomment-683793230), the issue seems to only appear when the application is built in release mode.

I would like to report the result of a simple internal test we did on our application during this weekend.
In our application we have a HealthKit observer query that is registered to execute in background. The setup of the query is something like the following:
```swift
let store = HKHealthStore()
let query = HKObserverQuery(sampleType: sampleType, predicate: nil, updateHandler: self.updateHandler)
store.execute(query)
store.enableBackgroundDelivery(for: sampleType, frequency: .hourly) { (succeeded, error) in
    // log..
}
```

We tested a build with this query commented out and this issue never occured.
Of course this is just a temporary solution for us, while we wait for this bug to be addressed.

**Comment by indigothm** [comment]
@ezamagni Same issue occurs for us when using Healthkit background observers

**Comment by indigothm** [comment]
More examples of the same issue in both my app and AboutYou online store and my own app:

![IMG_4178](https://user-images.githubusercontent.com/6455846/93523659-0b3b2180-f93c-11ea-894e-b41fe1d9cbb5.PNG)
![IMG_4170](https://user-images.githubusercontent.com/6455846/93523683-15f5b680-f93c-11ea-8e12-3048713b19ed.PNG)

It doesn't seem to be related to healthkit after all and seems to affect a wider range of apps

**Comment by mahdi-ninja** [comment]
Hey guys,

Any updates on this?

@dnfield @chinmaygarde

**Comment by pawlowskim** [comment]
The same issue in our application, but... I have a test case that works. We have background geofence that possibly triggers UI rendering. But this is not that relevant I think, because this happens for us when app is killed and then started. We are using https://github.com/transistorsoft/flutter_background_geolocation for geofence, and here is how we can reproduce the issue:

- kill the app
- use any software to simulate GPS location (fe. "iTools Pro")
- simulate location far away from the geofence (leave geofence range)
- simulate location couple times in geofence
- tap the application on the screen

We have custom loading screen when app is starting, which is always visible when we start application after killing it. But when we manipulate GPS while app is killed, and then we click the icon, it starts with the "dashboard" screen, instead of loading screen. It looks like it was run in the background somehow.

Moreover, our client says it triggers when app is in background and push notifications is send.

```
[✓] Flutter (Channel stable, 1.20.3, on Mac OS X 10.15.3 19D76, locale en-GB)

[!] Android toolchain - develop for Android devices (Android SDK version 28.0.3)
    ! Some Android licenses not accepted.  To resolve this, run: flutter doctor --android-licenses
[✓] Xcode - develop for iOS and macOS (Xcode 11.6)
[!] Android Studio (version 3.3)
    ✗ Flutter plugin not installed; this adds Flutter specific functionality.
    ✗ Dart plugin not installed; this adds Dart specific functionality.
[✓] IntelliJ IDEA Ultimate Edition (version 2020.1.3)
[✓] Connected device (3 available)
```
Tested on iPhone X iOS 14.0, iPhone 6 iOS 12.4 and iPhone 8 iOS 13.6.1. No matter if application is installed through TestFlight or debug version installed locally. Previously we were using Flutter 1.12.13+hotfix.8 and it was working (pre Metal age 🤘 ). We are upgrading to fluter 1.22 to see if that helps and also downgrading to 1.12.13+hotfix.8 so we can confirm it's working.

**Comment by pawlowskim** [comment]
I can also confirm, that when application is run by background task, views are rendered under the hood. Test case:
- run geofence tracking in the background (location permissions set to always allow), including run on device boot
- restart the device to be sure everything is freshly started
- force background task to start the application
- keep it running for the condition it should show new view
- open application
- no loading screens, views that are triggered by some events can be seen immediately

Our application count users steps, send it the backend and do some calculations with some time interval. When user met certain conditions, it pushes new view. So, after restarting the device application is run in the background (in detached state) and start working as it should. It tracks user GPS, collect steps, send to backend etc. When condition is met and specific view in the app should be shown I have started the application. This view is shown right away, even haptic feedback that is added to the view is working.

I can also confirm that checking if app is in detached state or not "fixes" the issue. But there is another problem. App state observer doesn't trigger that app is active when we run the application that was run in the background (https://api.flutter.dev/flutter/widgets/WidgetsBindingObserver-class.html). So simply adding if condition works when app is in detached state and is opened, but doesn't work when app isn't started from background.
We are testing getting this event directly from the iOS so we can receive each single event related with app state. This would be obviously workaround, but it may help track down the problems on flutter side. Cheers :)

**Comment by fluttercorner** [comment]
[How to set Background Image to Scaffold in Flutter](https://fluttercorner.com/how-to-set-background-image-to-scaffold-in-flutte/)

**Comment by mahdi-ninja** [comment]
@TahaTesser Hey there,

Do you know if there are any plans to work on this issue?

**Comment by pawlowskim** [comment]
We have upgraded flutter to 1.22.0 and also did a workaround with checking app state from native site to get each event. It seems to resolves the problems for now.

**Comment by dnfield** [comment]
I'd like to look at this, but I'm hoping there's some reproduction that is consistent and does not require hours (or minutes) of runtime to happen. Has anyone been able to consistently reproduce this issue with clear, determiniate steps?

**Comment by pawlowskim** [comment]
I guess, any code that starts your app in the background (when it's killed) will trigger it.

**Comment by indigothm** [comment]
@pawlowskim "also did a workaround with checking app state from native site to get each event" - I understand that you check the state using the observers, but what exactly do you do?

**Comment by pawlowskim** [comment]
Currently we are using (I know, very popular...) https://github.com/ened/flutter_native_app_state_plugin.git to check the UIApplication.State. In our loading page, if it at least once came as 0 (https://developer.apple.com/documentation/uikit/uiapplication/state/active) we allow to render our application, otherwise we keep user on loading page. That allow (at least in our case) to not render application views until it's attached to the view, so in worst case, user won't see some images on app loading page.

**Comment by dnfield** [comment]
The description in https://github.com/flutter/flutter/issues/64558#issuecomment-703648866 makes me believe something is prematurely causing the view to load. We try to be careful about this but it sounds like it may still be happening. 

We could try to be _extra_ defensive about this by making the metal surface impl refuse to do metal work until we've received applicationDidBecomeActive, but really we need to make sure we're not loading the view prematurely too.

**Comment by dnfield** [comment]
It would be super helpful if anyone who is able to reproduce this and compile a local engine were to instrument the FlutterView loading logic to print a backtrace and see what's causing it to load.

**Comment by igorczapski** [comment]
It's quite easy to reproduce when you know how to do it. 
1. Have some images on the first page you'll see after opening.(we're displaying images from the assets as well as CachedNetworkImage). Kill the app and run something that will wake up the app.
2. Trigger engine start in the background. It probably might be triggered by bluetooth (I've seen somewhere issue related to that) or, as in our case, an entrance to a geofence (we use flutter_background_geolocation package for that - and in xcode background modes there are selected following options: location updates, remote notifications, background fetch). Yes, the image is loaded prematurely IF and only IF flutter engine starts in background. 
3. Open console where you can read logs. There you can tell that engine is starting.
4. Open the app. Images are broken.

The app behaves as it would normally be in foreground, renders widgets etc. For instance, we have some widget which animates and has vibration tied to that animation. Animation starts in initState and it normally vibrates in the pattern dictated by that animation. And after flutter update it started vibrating in background. This is unwanted behaviour and happened in production app. When I open the app, I see the page which is already in the state when animation is finished, instead of seeing the views animate. 
When we were using pre-metal flutter version, there were no problems. I don't know what else to add. I've never compiled flutter engine, so I can't help with that, sorry.

**Comment by dnfield** [comment]
@igorczapski - can you please provide an application that reproduces this as you're explaining?

I tried for an hour or so the other day to get something working that did something close to that, and it wasn't quite happening.

**Comment by pawlowskim** [comment]
I have prepared an application that show the problem: https://github.com/pawlowskim/headless_mode_images_issue
The easiest way to trigger the problem is too:
* install the application
* kill it
* open console app and select your device to see the logs
* use iTools Pro to spoof the GPS (it will trigger location service to run the application)
* wait until there is `Build` printed in the console
* run the application

If the app is killed, and you start walking to trigger geo service app will be run and view will be build in the background. That causes problem with images. If you uncomment more images, it will even crash. The delay in the main is added to easier trigger the issue. Without it, there is less chance to see that images are not visible.
@dnfield FYI.

**Comment by BRZErez** [comment]
We are Having the same IOS issue, With a twist.
For us this started when moving from Flutter 1.16 to 1.17.
Then the rendering issues started on IOS.
It was sporadically and could be images or even Text, Just like a printer that needs to clean the printing nozzle (Text cut in the middle of letter and a letter beside it was printer good and so on.
We solved it buy moving to 1.19 and that seems to solve the issue... Now we moved to 1.20 and it is back and also on 1.22
We see it way more on IOS 14 then iOS 13 and it seems like it is only images now and not the Text anymore. But still, This is a production app and this is sooooo Bad!
And it is random in our case, Reproducing it is so hard... Always related the BG and returning issues but not a stable issue.

**Comment by Mehta2015** [comment]
Even we are facing the same issue when we deal with Health kit background steps sync.
Flutter sdk : 1.20.3
Any solution on it?

**Comment by pawlowskim** [comment]
@Mehta2015 so far only working solution for us is https://github.com/flutter/flutter/issues/64558#issuecomment-712773782

**Comment by Mehta2015** [comment]
> 
> 
> @Mehta2015 so far only working solution for us is [#64558 (comment)](https://github.com/flutter/flutter/issues/64558#issuecomment-712773782)

@pawlowskim  we tried but it's not working in our case. But still will check what you have suggested.

**Comment by indigothm** [comment]
@Mehta2015 Working solution - https://github.com/acoutts/flutter-engines-no-metal

**Comment by BRZErez** [comment]
We found a workaround or a solution....
I've read many articles about using SVGs in Flutter apps instead of PNGs.
I know that in IOS we hate SVGS and they just kill the app.
But when rendering flutter all the suggestion are to ditch PNGs and use SVGS...
Doing that for all images in our main screen solved the issue. Now more rendering issue no matter what. Working perfectly...
Give it a try and let us know... Worked perfectly, really wierd but and really bothering on how flutter works but as long as it works...

**Comment by rohanj1132** [comment]
Any leads on how this can be fixed?

**Comment by indigothm** [comment]
@rohanj1132 Try to use the Open GL version of Flutter engine I mentioned above. Literary need to replace one folder in the project

**Comment by knaeckeKami** [comment]
Just got a screenshot from a friendly user that looks like this:
![Images missing](https://i.imgur.com/rYYiZBA.jpg)

Flutter version: 1.22.4

All the images missing, some of them could be blamed on network issues, but at least the icons in the navigation bar are static assets.

I'm not aware of anything that could trigger an engine start in the background for this app.

Switching to OpenGL is not an option for us, there are other issues with OpenGL.

**Comment by gmWheeler001** [comment]
still getting this on 1.23.0
every time the app is resumed the images do not load correctly, some images work most dont. mostly users on iPhone X and later

**Comment by BRZErez** [comment]
The solution is to stop using PNGs in the filter side... on the flutter side use one SVGs and on the native side use PNGs. This works perfect as a fix.

**Comment by xvrh** [comment]
@pawlowskim 
Can you explain in more detail (a full example would be great) your solution with `waitUntilFirstFrameRasterized`?
This is a release blocker for me and I'm looking for any workaround

**Comment by mahdi-ninja** [comment]
@dnfield Did you see the sample project that @pawlowskim prepared? https://github.com/flutter/flutter/issues/64558#issuecomment-717834560

It's been more than 4 months and this ticket still has not been resolved :(

**Comment by pawlowskim** [comment]
> @pawlowskim
> Can you explain in more detail (a full example would be great) your solution with `waitUntilFirstFrameRasterized`?
> This is a release blocker for me and I'm looking for any workaround

We have wrapped each call to navigation with `waitUntilFirstFrameRasterized` so, we don't push new page until first frame was rasterized (displayed on screen). The worst case is that first possible widget you want to display may not work (in our app it's loading page, so it's not a big deal).

```
final GlobalKey<NavigatorState> navigatorKey = GlobalKey<NavigatorState>();

MaterialApp(
                title: 'My App',
                navigatorKey: navigatorKey,
                theme: ThemeCreator.mainThemeData,
                initialRoute: initialRoot,
                routes: routes,
                builder: appBuilder)
```

and when you want to push any route we use something like
```
  return WidgetsBinding.instance.waitUntilFirstFrameRasterized.then((value) {
    return navigatorKey.currentState.push(MaterialPageRoute<T>(builder: builder, settings: settings));
  });
```

**Comment by SaumilPatelGreenApex** [comment]
Getting the same issue when the IOS application starts after tapping on the custom notification, displayed from FirebaseMessaging.onBackgroundMessage() when App is killed.

**Comment by SaumilPatel1906** [comment]
Getting the same issue when the IOS application starts after tapping on the custom notification, displayed from FirebaseMessaging.onBackgroundMessage() when App is killed.

**Comment by msangals** [comment]
We faces the same issue, here an example video:

Image rendering is working as expected:

https://user-images.githubusercontent.com/4263095/109479904-31ccbf00-7a7b-11eb-8091-cdab4d044707.MOV

Image rendering not works anymore after starting app from background after a couple of time

https://user-images.githubusercontent.com/4263095/109479926-372a0980-7a7b-11eb-9aab-563c45d9156e.MOV

<details>
<summary>flutter doctor -v</summary>

```
flutter doctor -v
[✓] Flutter (Channel stable, 1.22.6, on macOS 11.2 20D64 darwin-x64, locale en-GB)
    • Flutter version 1.22.6 at /Users/msangals/flutter
    • Framework revision 9b2d32b605 (5 weeks ago), 2021-01-22 14:36:39 -0800
    • Engine revision 2f0af37152
    • Dart version 2.10.5

[!] Android toolchain - develop for Android devices (Android SDK version 30.0.0-rc1)
    • Android SDK at /Users/msangals/Library/Android/sdk
    • Platform android-30, build-tools 30.0.0-rc1
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)
    ✗ Android license status unknown.
      Run `flutter doctor --android-licenses` to accept the SDK licenses.
      See https://flutter.dev/docs/get-started/install/macos#android-setup for more details.

[✓] Xcode - develop for iOS and macOS (Xcode 12.4)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.4, Build version 12D4e
    • CocoaPods version 1.10.1

[✓] Android Studio (version 4.0)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 46.0.2
    • Dart plugin version 193.7361
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)

[✓] IntelliJ IDEA Ultimate Edition (version 2020.3.2)
    • IntelliJ at /Applications/IntelliJ IDEA.app
    • Flutter plugin installed
    • Dart plugin version 203.6912

[!] Connected device
    ! No devices available

! Doctor found issues in 2 categories.
```
</details>

**Comment by BRZErez** [comment]
Did anyone tried the solution i added few  month ago? No one reacted but for us it made all the deference and we don't have any images or artefacts issues anymore.
I wonder why no one tried it?
Stop using PNGs, Just use only SVGs and it should solve the issue.
If you come from an IOS background it seems very weird as Native iOS doesn't like SVGs because there is a need to render them all the time. But the fact it that in Flutter engine it is working way better than PNGs.
So what I'm doing is to use SVGs when something is rendered in a Flutter canvas and PNGs when something is written directly for iOS native.
I hope this helped as I see that people keep asking for a solution for the issue and it there very simple until the flutter stuff will fix it over there (I'm not optimistic they will).

**Comment by rohanj1132** [comment]
@BRZErez do you use the flutter_svg package to render the SVG?

**Comment by knaeckeKami** [comment]
SVGs are much more expensive to render and even natively supported by flutter. It might be a workaround, but it's not a solution

**Comment by rohanj1132** [comment]
@knaeckeKami I agree, you cannot even replace all types of images with SVG 
but for certain situation like the bottom navigation where we are using custom images I can replace by SVG.

**Comment by BRZErez** [comment]
> @BRZErez do you use the flutter_svg package to render the SVG?

Yes

**Comment by BRZErez** [comment]
> @knaeckeKami I agree, you cannot even replace all types of images with SVG
> but for certain situation like the bottom navigation where we are using custom images I can replace by SVG.

You are right and I wrote in my original answer the same thing. I do not want it to be a permanent solution. But I found it to work and the Flutter team support it as I found someone from the team that say that in flutter it is better to use SVG so it support my claim.
I have many issues with Flutter, It is not the only one. When going to bigger more advance projects with lots of native needs Flutter seems to be less and less desired but the PM didn't ask about that. They asked about how to solve the images that are not rendering and this is a solution. 
There is no really better solution right now. If you find one please let me know ( One that always work ) and I would love to implement that.

**Comment by BRZErez** [comment]
> @knaeckeKami I agree, you cannot even replace all types of images with SVG
> but for certain situation like the bottom navigation where we are using custom images I can replace by SVG.

I believe that you need to replace only the images that are being displayed in the main screen that comes up when the app is loaded., I don't think that you need to change any images in inner widgets that are not being displayed when the app is being loaded.
That is what I think.
But make sure to replace all images in the tree hierarchy for the main screen (That is why I am writing screen and not widget, So there won't be any confusion.

**Comment by gaaclarke** [comment]
This may be the same issue as https://github.com/flutter/flutter/issues/71074, it's just that on some devices the image is transparent and on others it is magenta (pink).

**Comment by rohanj1132** [comment]
Hi @gaaclarke, I saw your comments on #71074 I will test the fix and report back. It takes sometime to reproduce the issue, so I will report back in 2 days.

**Comment by MirceaX2Mobile** [comment]
I'm having the same problems on iOS, I tried every caching framework, but nothing works not even the flutter default network image caching. After the app resumes all the images are fetched again. Any fixes on this?

**Comment by rohanj1132** [comment]
@gaaclarke Tested the fix on iphone 12 (me) and iphone 11 (QA in my team).
The issue was still reproducible on iphone 11. The missing images are not pink but rather transparent/invisible.
On iphone 12, I have not been able to reproduce the issue so far (but given the nature of the issue, I cannot be confident that issue is fixed).

Adding an image as well (ignore the red box, just added it for privacy)
![images_not_loading](https://user-images.githubusercontent.com/6575512/110899996-0d4cce80-8328-11eb-944d-b94eb113f4b4.png)

**Comment by gaaclarke** [comment]
@rohanj1132 How did you test the fix?

This demonstratively fixes the issue if you use the steps from https://github.com/flutter/flutter/issues/64558#issuecomment-717834560

If you can provide reproduction code of your issue, could you please file a different issue?  Maybe you are running into something else.

**Comment by gaaclarke** [comment]
@rohanj1132 I'm just realizing you may have been testing https://github.com/flutter/engine/pull/24503, not https://github.com/flutter/engine/pull/24958.

To clear up any confusion, a few weeks ago I made a speculative fix since I couldn't reproduce the bug, that was https://github.com/flutter/engine/pull/24503, that's on master currently.   Yesterday I was finally able to reproduce the bug and I had to tweak the fix from a few weeks ago ( https://github.com/flutter/engine/pull/24958 ).  The latest PR has not landed, it is still in review.

As part of landing that PR they might ask me to revert the previous change and then commit them again together as one commit, we'll see.

**Comment by rohanj1132** [comment]
@gaaclarke Yes. Eagerly awaiting the new PR to be merged. I will test it out once it lands on master.
Thank you for looking into this issue.

**Comment by abhaysood** [comment]
@TahaTesser Could you tell the version in which this was fixed?

**Comment by TahaTesser** [comment]
@abhaysood 
Please try `master` channel

**Comment by onikiri2007** [comment]
Hi, When will this fix be in beta or stable?

**Comment by gaaclarke** [comment]
@onikiri2007 I don't know when it will be on stable.  You can track it's progress here (https://github.com/flutter/flutter/commit/4266d629c26213ac5cd3f51c875f84e408d33a14).  I talked to the team about potentially including this in a hotfix, the problem is that because we couldn't create an automated test for it there was apprehension to cherrypick it.  That means it will follow the standard processes of Master -> Beta -> Stable.

**Comment by MirceaX2Mobile** [comment]
I tested today from Stable and this issue is not fixed. 
I think this is related to Caching logic on iOS. If you put the app in background the cache is automatically cleared, I don't know why.

Does anyone have any idea why this is happening?

**Comment by gaaclarke** [comment]
@MirceaX2Mobile check out my last comment.

**Comment by dnfield** [comment]
This commit should be available in the latest stable. If you're still seeing the issue you may have a different bug.

The image cache is cleared when going to background to conserve memory and try to avoid the OS being as likely to kill the app. It should just re-decode the image when you come back and see a cache miss. Please feel free to file a new bug if you're still seeing a similar issue though.

**Comment by gaaclarke** [comment]
Ahh yea, looks like 2.2.0 has it, awesome.  That came out on 5/18/2021.

**Comment by MirceaX2Mobile** [comment]
@gaaclarke Yup it’s up and running so the bug with the magenta placeholder should be fixed.

@dnfield Thank you for the info. On android the cache works even if you put the app in background. Do you know if there is any handling / workaround to make this work on iOS too?

**Comment by dnfield** [comment]
The cache should work equally well on both platforms, and should get cleared when the app backgrounds on both platforms.

Please file a new bug with a minimal reproduction of this, and instructions about how to reproduce it. Feel free to ping me on it.

**Comment by MirceaX2Mobile** [comment]
@dnfield 
Thank you! I will try tommorow to show you a demo about this. Thanks again for the support!

**Comment by MirceaX2Mobile** [comment]
@dnfield So as I said, the image cache remains on android even if you put the app in background. I attached a demo below android vs iOS. I am using a PageView and for loading the images I'm using FadeInImage.assetNetwork. Any ideas why this is happening?

https://user-images.githubusercontent.com/33491440/119219905-7d2dc100-baf0-11eb-98a9-cab44b21e8da.mov

https://user-images.githubusercontent.com/33491440/119219909-7ef78480-baf0-11eb-8f5c-a04510d83790.mov

**Comment by MirceaX2Mobile** [comment]
Does anyone have any explanations for the above behaviour?

**Comment by MirceaX2Mobile** [comment]
@dnfield
So you are saying that this should be a reportable bug, right? I would really like to not clear the cache on my use case on iOS.

**Comment by dnfield** [comment]
Ok, rereading the source some more:

- On Android, we make a call to notify the Dart side that memory is running low only under certain conditions. These conditions may or may not be triggered when going to background.
- On iOS, we make this call whenever we lose the view controller, go to the background, or get a low memory warning. This follows practices recommended by Apple - applications should not hold on to resources when in background that can be recovered, particularly graphics related resources. i-devices tend to have less memory as well, and you are more likely to get killed if you're retaining too much memory in the background.

There's currently no way to override this behavior - the painting binding implementation does it unconditionally when it gets the handle memory pressure warning.  You could, however, precache images when you resume from background.

IMHO, the bug is that we're not doing this as often on Android.

**Comment by MirceaX2Mobile** [comment]
@dnfield Thank you so much for the detailed answer. 
“You could, however, precache images when you resume from background.” You mean to make a custom cache by myself , wouldn’t be cleared too?

**Comment by dnfield** [comment]
I mean using `precacheImage` when your application resumes, but yes using a custom cache would also work (though you'd have some work to do to make it play nicely with the rest of the image subsystem)
